### PR TITLE
23 mailing abilities

### PR DESCRIPTION
**Pull request by Shubham2301**
PR for #23

### CODE REVIEW

File: public/wp-content/themes/ColoredCow/functions.php [hunk 1]
```diff
@@ -49,6 +49,8 @@ function add_subscriber(){
             add_post_meta($post_id, 'password', $hash);
             add_post_meta($post_id, 'gender', $post_gender);
         }
+        require "mailer.php";
```
Comment (rathorevaibhav) [inline public/wp-content/themes/ColoredCow/functions.php]: use require_once.

File: public/wp-content/themes/ColoredCow/mailer.php [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+                        array(
+                            'name' => 'first_name',
+                            'content' => $guest_name
+                        )
+                    )
+                )
+            ),
+        );
+        $result = $mandrill->messages->sendTemplate($template_name, $template_content, $message);
+        // var_dump($result);
```
Comment (rathorevaibhav) [inline public/wp-content/themes/ColoredCow/mailer.php]: remove comments please.

File: public/wp-content/themes/ColoredCow/functions.php [hunk 3]
```diff
@@ -49,6 +49,8 @@ function add_subscriber(){
...
+        register_email($post_email,$post_title,'shubham@coloredcow.com','Shubham');
```
Comment (rathorevaibhav) [inline public/wp-content/themes/ColoredCow/functions.php]: host email and name should be dynamic. Please use 'admin' user's credentials.

File: public/wp-content/themes/ColoredCow/mailer.php [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        throw $e;
+    }
+}
+
+function rsvp_email($guest_email,$guest_name,$host_email,$host_name,$soiree_date,$soiree_name)
+{   
+    try {
+        $mandrill = new Mandrill('v0tqtpCwhDCIOLFe5Hw-gA');
+        $template_name = 'ThanksForRSVP';
+        // $template_content = array(
```
Comment (rathorevaibhav) [inline public/wp-content/themes/ColoredCow/mailer.php]: remove comments.

File: public/wp-content/themes/ColoredCow/mailer.php [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        echo 'A mandrill error occurred: ' . get_class($e) . ' - ' . $e->getMessage();
+        // A mandrill error occurred: Mandrill_Unknown_Subaccount - No subaccount exists with the id 'customer-123'
+        throw $e;
+    }
+}
+
+function rsvp_email($guest_email,$guest_name,$host_email,$host_name,$soiree_date,$soiree_name)
+{   
+    try {
+        $mandrill = new Mandrill('v0tqtpCwhDCIOLFe5Hw-gA');
```
Comment (rathorevaibhav) [inline public/wp-content/themes/ColoredCow/mailer.php]: This is a secret key and should be in wp-config. Should be manually entered for each environment. Please remove it from here and use PHP global variables in wp-config.

File: public/wp-content/themes/ColoredCow/mailer.php [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+                        array(
+                            'name' => 'soiree_date',
+                            'content' => $soiree_date
+                        ),
+                    )
+                )
+            ),
+        );
+        $result = $mandrill->messages->sendTemplate($template_name, $template_content, $message);
+        // var_dump($result);
```
Comment (rathorevaibhav) [inline public/wp-content/themes/ColoredCow/mailer.php]: remove all comments.

File: public/wp-content/themes/ColoredCow/mailer.php [hunk 7]
```diff
@@ -0,0 +1,105 @@
+<?php
+/**
+ * Template Name: Mailer
+ */
+
+require_once get_template_directory() . '/vendor/mandrill/mandrill/src/Mandrill.php';
+function register_email($guest_email,$guest_name,$host_email,$host_name)
+{   
+    try {
+        $mandrill = new Mandrill('v0tqtpCwhDCIOLFe5Hw-gA');
```
Comment (rathorevaibhav) [inline public/wp-content/themes/ColoredCow/mailer.php]: remove key from here.

File: public/wp-content/themes/ColoredCow/mailer.php [hunk 8]
```diff
@@ -0,0 +1,105 @@
+<?php
+/**
+ * Template Name: Mailer
+ */
+
+require_once get_template_directory() . '/vendor/mandrill/mandrill/src/Mandrill.php';
+function register_email($guest_email,$guest_name,$host_email,$host_name)
```
Comment (rathorevaibhav) [inline public/wp-content/themes/ColoredCow/mailer.php]: We can use a Wrapper class for Mandrill. I see a lot of code repetition in both functions. Can you read about creating a wrapper class and class's constructor?

**Review by rathorevaibhav — CHANGES_REQUESTED** [review]

**Comment by rathorevaibhav** [comment]
Testing looks good. Onto Code Review.

File: public/wp-content/themes/ColoredCow/composer.json [hunk 9]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (rathorevaibhav) [inline public/wp-content/themes/ColoredCow/composer.json]: is this file needed now? since you are using wpMandrill.

File: public/wp-content/themes/ColoredCow/functions.php [hunk 10]
```diff
@@ -30,28 +31,49 @@ function cc_styles() {
...
+            wp_send_json_error();
         }
+
+        add_post_meta($post_id, 'phone', $post_phone);
+        add_post_meta($post_id, 'email', $post_email);
+        add_post_meta($post_id, 'password', $hash);
+        add_post_meta($post_id, 'gender', $post_gender);
+        $host_info = get_userdata(get_current_user_id());
+        $host_name = $host_info->first_name;
+        // var_dump($host_name);
```
Comment (rathorevaibhav) [inline public/wp-content/themes/ColoredCow/functions.php]: remove comments please.

**Comment by rathorevaibhav** [comment]
Confirm and Decline button not working on Admin dashboard.
image

**Comment by rathorevaibhav** [comment]
Mails from different users. Should be consistent for everyone. The first one came when I was logged in. The second one came when I wasn't.
image

File: public/wp-content/themes/ColoredCow/functions.php [hunk 11]
```diff
@@ -67,22 +89,40 @@ function verify_credentials(){
...
         $rsvp_date = date("d/m/y");
-        $rsvp_guest_id = $wpdb->get_var("SELECT post_id FROM $tablename WHERE meta_value = '".$_POST['guest_email']."'");
+        $guest_email = $_POST['guest_email'];
+        $rsvp_guest_id = $wpdb->get_var("SELECT post_id FROM $tablename WHERE meta_value = '".$guest_email."'");
         $hash = get_post_meta($rsvp_guest_id,'password',true);
-
+        $guest_name = get_the_title($rsvp_guest_id);
+        $soiree_name = get_the_title($event_id);
+        $soiree_date = get_post_meta($event_id,'event_date',true);
+        $host_info = get_userdata(get_current_user_id());
```
Comment (rathorevaibhav) [inline public/wp-content/themes/ColoredCow/functions.php]: This might not work if the current user doesn't have an account. Issue reported as testing feedback. It'll produce different results when current is signed/not signed.

File: public/wp-content/themes/ColoredCow/functions.php [hunk 12]
```diff
@@ -67,22 +89,40 @@ function verify_credentials(){
...
             wp_send_json_error("failed");
         }
         if(array_key_exists($rsvp_guest_id,$event_users)){
             wp_send_json_error("duplicate");
         }
         $event_users = get_post_meta( $event_id, $meta_key, true );
-        $guest_name = get_the_title( $rsvp_guest_id );
         $event_users[$rsvp_guest_id] = get_rsvp_guest_meta( $guest_name );
         $event_users ? update_post_meta( $event_id, $meta_key, $event_users ) : add_post_meta($event_id, $meta_key, $event_users);
+        // rsvp mail
```
Comment (rathorevaibhav) [inline public/wp-content/themes/ColoredCow/functions.php]: remove comment.

File: public/wp-content/themes/ColoredCow/functions.php [hunk 13]
```diff
@@ -67,22 +89,40 @@ function verify_credentials(){
...
+        $guest_email = $_POST['guest_email'];
+        $rsvp_guest_id = $wpdb->get_var("SELECT post_id FROM $tablename WHERE meta_value = '".$guest_email."'");
         $hash = get_post_meta($rsvp_guest_id,'password',true);
-
+        $guest_name = get_the_title($rsvp_guest_id);
+        $soiree_name = get_the_title($event_id);
+        $soiree_date = get_post_meta($event_id,'event_date',true);
+        $host_info = get_userdata(get_current_user_id());
+        $host_name = $host_info->first_name;
+        $host_email = get_option('admin_email');
```
Comment (rathorevaibhav) [inline public/wp-content/themes/ColoredCow/functions.php]: Are we using this variable? Check line [114](https://github.com/Shubham2301/phase2/pull/33/files#diff-01391f7f399cd5492a25445b0fd9ec87R114).

File: public/wp-content/themes/ColoredCow/mailer.php [hunk 14]
```diff
@@ -0,0 +1,104 @@
+<?php
+
+require_once __DIR__ . '/../../plugins/wpmandrill/lib/mandrill.class.php';
```
Comment (rathorevaibhav) [inline public/wp-content/themes/ColoredCow/mailer.php:3]: use plugins directory path. There should be an inbuilt WP function to get the plugins directory.

File: public/wp-content/themes/ColoredCow/mailer.php [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+    function set_template( $template )
+    {
+        $this->template = $template;
+    }
+
+    function set_merge_vars( $merge_vars )
+    {
+        $formatted_merge_vars = array();
+        foreach( $merge_vars as $rcpt => $merge_var )
+            {
```
Comment (rathorevaibhav) [inline public/wp-content/themes/ColoredCow/mailer.php]: fix indentation on this one and the next line.

**Comment by rathorevaibhav** [comment]
@Shubham2301 one more feedback: I cannot access individual soiree page. Are we tracking that in a different PR or was already done? The only Soiree I can register is the latest one created.

File: public/wp-content/themes/ColoredCow/single-soiree.php [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+								<div class="col-md-8 rsvp text">
+									<?php
+										$page = get_page_by_title("Register Guest");
+										$link_register_page = get_page_link($page->ID) . "?eid=" . $post->ID;
+									?>
+									<p class="row link text-center"> If not! then hurry up and become a guest user by:</p>
+									 <a class="btn btn-primary btn-lg" role="button" href="<?php echo $link_register_page;?>" target="_blank">REGISTERING HERE</a>
+								</div>
+							</div>
+							<?php echo do_shortcode('[addtoany]'); ?>
```
Comment (rathorevaibhav) [inline public/wp-content/themes/ColoredCow/single-soiree.php]: what is this for?

**Review by rathorevaibhav — CHANGES_REQUESTED** [review]
Have a look at the requested changes.

**Comment by Shubham2301** [comment]
single soiree page is done in another issue.
